src/java/test/org/apache/zookeeper/server/NIOServerCnxnTest.java

```diff
@@ -68,4 +74,38 @@ public void testOperationsAfterCnxnClose() throws IOException,
             zk.close();
         }
     }
+
+    /**
+     * Mock extension of NIOServerCnxn to test for
+     * CancelledKeyException (ZOOKEEPER-2044).
+     */
+    class MockNIOServerCnxn extends NIOServerCnxn {
+        public MockNIOServerCnxn( NIOServerCnxn cnxn )
+                throws IOException {
+            super( cnxn.zkServer, cnxn.sock, cnxn.sk, cnxn.factory );
+        }
+
+        public void mockSendBuffer(ByteBuffer bb) throws Exception {
+            super.internalSendBuffer( bb );
+        }
+    }
+
+    @Test(timeout = 30000)
+    public void testValidSelectionKey() throws Exception {
+        final ZooKeeper zk = createClient();
+        try {
+            Iterable<ServerCnxn> connections = serverFactory.getConnections();
+            for (ServerCnxn serverCnxn : connections) {
+                MockNIOServerCnxn mock = new MockNIOServerCnxn((NIOServerCnxn) serverCnxn);
+                // Cancel key
+                ((NIOServerCnxn) serverCnxn).sock.keyFor(((NIOServerCnxnFactory) serverFactory).selector).cancel();;
+                mock.mockSendBuffer( ByteBuffer.allocate(8) );
+            }
+        } catch (CancelledKeyException e) {
+            e.printStackTrace();
```

Comment:
Instead of printStackTrace, can we use LOG.error("Exception while sending bytes!", e);